### DISCUSSION TITLE
Figure out how to support other cloud providers than AWS, GCP or Azure

### DISCUSSION THREAD

**Discussion by mrchrisadams** [cloud-carbon-footprint/cloud-carbon-footprint#1041] [Ideas]
## Feature Suggestion
**Is your feature request related to a problem? Please describe.**

I want to quantify and manage the carbon emissions from infrastructure I operate, but the provider is not Amazon AWS, Microsoft Azure, or Google Cloud Platform.

**Describe the solution you'd like**

If I know the same info that is used for current calculations, like instance size, time of use, geographic location etc, then I would ideally be able to make the same kinds of calculations for other providers.

We use [Hetzner](https://www.hetzner.coml) and [Scaleway](https://www.scaleway.com/en/) at the Green Web Foundation, and both providers expose APIs to provide hourly billing, as well as machine readable descriptions of the services they expose.

You can see an example below of the info Hetzner exposes. They provide listings for server types, along with prices like the other providers do:
https://docs.hetzner.cloud/#server-types

**Describe alternatives you've considered**

I've investigated, [Scaphandre][1], and based on what I learned, I [contributed this section to the docs explaining how it works][2], before realising that while it probably _could_ be used for this, it was a bit low level, and I would need to use some observability tool like Grafana/Prometheus to expose the metrics.

I haven't discounted that, but building all the logging, infra is a massive job, and I figured consuming data from an existing cost/usage API might be a more manageable first step.

[1]: https://github.com/hubblo-org/scaphandre
[2]: https://hubblo-org.github.io/scaphandre-documentation/explanations/how-scaph-computes-per-process-power-consumption.html

**Additional context**

## Possible Implementation

Hetzner expose some usage metrics to customers anyway (see an example below form one machine).

I also know enough about the specs of the machine I'm using to make to come up with some sample numbers for:

- likely energy use per hour for the VMs I am using
- grid intensity figures for Norway and Germany

I don't know what the data needs to look like though to work to be consumed by cloud carbon footprint.

If I knew what the mocked data needed to look like and how to generate it, I figured it might be plausible to figure out the required 'shape' for data to allow various rollups and queries.

![Screenshot 2021-10-07 at 16 37 14](https://user-images.githubusercontent.com/17906/136406889-3023bd09-fbc2-4656-a2a3-7ac936c9da9e.png)

I also know that Scaphandre (the project I mentioned above) to offers per-process level energy usage readings based on RAPL and other readings. IT also supports the labelling of all these metrics so they can be scraped by tools like Prometheus, and others, to allow for all kinds of time series queries. This could plausibly support the same time based queries made by cloud carbon footprint.

I think Scaphandre already works with Kubernetes and Docker, so presumably if there's a format that can be specified, then basically any cloud provider can implement support for cloud carbon footprint too.

Wouldn't that be nice?

**Comment by ccasher** [comment]
Hi Chris, regarding the data needed to generate the estimates, you can use AWS as an example. In the [CostAndUsageReportsRow.ts](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/blob/trunk/packages/aws/src/lib/CostAndUsageReportsRow.ts) file, you'll notice on lines 38-46 we are defining the necessary variables needed for the energy estimation calculation. In order to define these variables, we are grabbing the vales needed from each row returned in the Athena query and you can see how we manipulate them in the class methods.

Depending on the usage unit, we are doing the energy calculation differently which you can see by the switch statement in [CostAndUsageReport.ts](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/blob/trunk/packages/aws/src/lib/CostAndUsageReports.ts). Once we calculate the energy, we just multiply the emissions factor to get co2e estimations.

Hopefully this helps. The other two cloud providers have similar logic in their respective packages.

**Comment by mrchrisadams** [comment]
ok, thanks @ccasher.

I've added the code here for discussion - lets assume I'm just using ephemeral cloud compute from Hetzner, just like I might be using ephemeral EC2 compute with AWS.

### Taking the billing based approach

If I take the billing based approach, I'd presumably need to be able to present my spending in as something that cloud carbon footprint can represent, I  need to find a way to take any billing data and make look/act a bit like a `BillingDataRow` you linked in [CostAndUsageReportsRow.ts](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/blob/trunk/packages/aws/src/lib/CostAndUsageReportsRow.ts#L22-L47),  right?

Maybe something like this pseudo code?

```ts
export default class HeznerCostAndUsageReportsRow extends BillingDataRow {
  constructor(usageRowsHeader: Hetzner.ValuesHeader, rowData: Hetzner.dat) {
    const billingDataRowKeys = usageRowsHeader.Data.map((column) =>
      Object.values(column),
    ).flat()
    const billingDataRowValues = rowData
      .map((column) => Object.values(column))
      .flat()
    const billingDataRow = Object.fromEntries(
      billingDataRowKeys.map((_, i) => [
        billingDataRowKeys[i],
        billingDataRowValues[i],
      ]),
    )
    super(billingDataRow)

    this.vCpuHours = this.getVCpuHours(Number(this.vCpus)) // how many CPUs I have running on this kind of instance
    this.usageAmount = this.getUsageAmount() // ie 400
    this.usageUnit = this.getUsageUnit() // i.e. CPU hours 
    this.timestamp = new Date(this.timestamp) // self explanatory
    this.cost = Number(this.cost) // something like cost-per-hour for this service or whatever is used
    this.accountId = this.accountName // self explanatory
    this.cloudProvider = 'Hetzner' // sub in the cloud provider here, i.e. Hetzner
    this.instanceType = this.parseInstanceTypeFromUsageType() // figure out the machine type from the billing Row
    this.replicationFactor = this.getReplicationFactor(billingDataRow) // if I'm using object storage, account for the number of times it's copied to account for the hardware providing the redundancy
  }
```

Taking the usage based approach

Alternatively, if I was going based on direct usage, like CPU usage figures, IOPs and the rest for machines, I might to this by pulling it from grafana/prometheus if I have that in place - a bit like the docs outline here when talking about using the cloud usage APIs themselves to come up with usage for figures for the amount of compute I'm using in a given time period.

https://www.cloudcarbonfootprint.org/docs/methodology#2-using-cloud-usage-apis-for-cloud-usage-higher-accuracy

Would you let me know which object I should look to see how that's implemented?

If there's a sample one for EC2, I think it might be possible to create a more generalised one for VMs from other providers.

In fact, because that Scaphandre supports [tracking individual qemu/kvm virtual machine energy usage](https://hubblo-org.github.io/scaphandre-documentation/how-to_guides/propagate-metrics-hypervisor-to-vm_qemu-kvm.html), and as well as tagging these at a service level (see the screenshot below), you could plausibly create objects to represent other services too, like how cloud carbon footprint seems to represent Redis compatible caching with Elasticache, and so on.

![image](https://user-images.githubusercontent.com/17906/137890317-0e4bc1d5-6a04-4d3a-8c7a-d83826c9085d.png)

If we know that the output needs to look like for cloud carbon footprint to parse it, I think it's possible to use prometheus/grafana to expose a feed and present similar kinds of charts.

**Comment by mrchrisadams** [comment]
OK, I've had a closer look at this [EC2.ts file here](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/blob/trunk/packages/aws/src/lib/EC2.ts).

It appears to be making its own queries to AWS Cost Explorer, for just EC2, to return a set of daily summaries of CPU hours, with given time period,.

I'm less familiar with what is done with the data when it's back, but if you know how many cloud instances you're running at given point in time, you could presumably create a daily summary the same way from Grafana/Prometheus tool.

If you're running 2 instances full time, you'd simply have two sets of 24 hours of CPU usage per day, presented as a daily rollup of 48hours for that day for example. And could even likely work out a mean utilisation figure too, or sub in a modelled number if it's not present.

**Comment by davidmytton** [comment]
Isn't the key difference that for AWS and GCP, you can pull the actual CPU utilization from the monitoring API whereas if you're working with billed hours you have to assume a mean CPU utilization? That's why the former is more accurate.

This is either pulled from a monitoring API observing the instance or from a billing report. 

What you're suggesting with Scaphandre allows the former type of reporting where there is no API because it is running on the instance and querying the OS for CPU stats. Presumably this is analagous to what the cloud monitoring APIs do (except they'll be doing it from the host/platform level). You probably don't want to average it though because otherwise it's no better than just assuming utilization from the billing hours.

**Comment by mrchrisadams** [comment]
Hi David 

> You probably don't want to average it though because otherwise it's no better than just assuming utilization from the billing hours.

That's a good point. I was being a bit lazy thinking it would be a pain to come up with a sensible figure for utilisation over the whole day, but I think the EC2 module works provides daily summaries based on the a rollup of the hourly mean CPU utilisation figures returned by the query.

I'm basing that [on this bit here in EC2.ts](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/blob/trunk/packages/aws/src/lib/EC2.ts#L81-L113) which I assume gives some ideas of what is returned for daily rollups:

```ts
  async getCosts(start: Date, end: Date, region: string): Promise<Cost[]> {
    const params: GetCostAndUsageRequest = {
      TimePeriod: {
        Start: start.toISOString().substr(0, 10),
        End: end.toISOString().substr(0, 10),
      },
      Filter: {
        And: [
          {
            Dimensions: {
              Key: 'REGION',
              Values: [region],
            },
          },
          {
            Dimensions: {
              Key: 'USAGE_TYPE_GROUP',
              Values: ['EC2: Running Hours'],
            },
          },
        ],
      },
      Granularity: 'DAILY',
      GroupBy: [
        {
          Key: 'USAGE_TYPE',
          Type: 'DIMENSION',
        },
      ],
      Metrics: ['AmortizedCost'],
    }
    return getCostFromCostExplorer(params, this.serviceWrapper)
```

And [this bit here in EC2.ts](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/blob/trunk/packages/aws/src/lib/EC2.ts#L39-L64), where presumably, that daily summary is created from a set of hourly averages for the day, assuming we have those figures:

```ts
private runQuery = async (
    start: Date,
    end: Date,
  ): Promise<ComputeUsage[]> => {
    const params = {
      StartTime: start,
      EndTime: end,
      MetricDataQueries: [
        {
          Id: 'cpuUtilizationWithEmptyValues',
          Expression:
            "SEARCH('{AWS/EC2,InstanceId} MetricName=\"CPUUtilization\"', 'Average', 3600)",
          ReturnData: false,
        },
        {
          Id: 'cpuUtilization',
          Expression: 'REMOVE_EMPTY(cpuUtilizationWithEmptyValues)',
        },
        {
          Id: 'vCPUs',
          Expression:
            'SEARCH(\'{AWS/Usage,Resource,Type,Service,Class } Resource="vCPU" MetricName="ResourceCount"\', \'Average\', 3600)',
        },
      ],
      ScanBy: 'TimestampAscending',
    }

```

If I have access to CPU usage as a metric, and and can do the same hourly metrics, I think you can come up with some similar numbers.

**Comment by mrchrisadams** [comment]
Oh neat! Hetzner totally has a metrics API endpoint, that can expose CPU usage, disk and network.

I think this means there's no need to even run your own metrics box, to get the same resolution as cloud carbon footprint defaults to for AWS and GCP.

https://docs.hetzner.cloud/#servers-get-metrics-for-a-server

Type | Timeseries | Unit | Description
-- | -- | -- | --
cpu | cpu | percent | Percent CPU usage
disk | disk.0.iops.read | iop/s | Number of read IO operations per second
  | disk.0.iops.write | iop/s | Number of write IO operations per second
  | disk.0.bandwidth.read | bytes/s | Bytes read per second
  | disk.0.bandwidth.write | bytes/s | Bytes written per second
network | network.0.pps.in | packets/s | Public Network interface packets per second received
  | network.0.pps.out | packets/s | Public Network interface packets per second sent
  | network.0.bandwidth.in | bytes/s | Public Network interface bytes/s received
  | network.0.bandwidth.out | bytes/s | Public Network interface bytes/s sent

You get a response like this. I've annotated it:

```js
{
  "metrics": {
    // should be self explanatory
    "end": "2017-01-01T23:00:00+00:00",
    "start": "2017-01-01T00:00:00+00:00",
    // this is the step, in seconds. I think we'd make this 3600 to get hourly averages like the EC2.ts approach
    "step": 60,
    // our time series that we'd consume to make daily summaries, to show
    "time_series": {
      "name_of_timeseries": {
        "values": [
          [
            1435781470.622,
            "42"
          ],
          [
            1435781471.622,
            "43"
          ]
        ]
      }
    }
  }
}

```

There are some caveats:

> Metrics are available for the last 30 days only. We limit the number of samples returned to a maximum of 500.

You'll want to be able persist metrics somewhere if you want to query further back than 30 days, and you'd need at least two queries to allow for a reading for every hour of a month (500 per API request, but 672 hours in a 4 week month).

Hetzner has two regions, in Germany and Finland respectively, so you can get the annualised CO2 intensity figures as well here.

I also think you'd also need one query per machine you're running, but you can get those ids, plus the machine's datacentre location with this API cll

https://docs.hetzner.cloud/#servers-get-all-servers

This actually seems fairly doable, at least for a Hacktoberfest experimental project.

**Comment by ccasher** [comment]
Hi @mrchrisadams, this looks great and totally doable. Please don't hesitate to reach out with any more questions and we can try to support this!

**Comment by mrchrisadams** [comment]
Thanks folks.

I'm gonna schedule sometime to work on this next week.

In the meantime, would you point me to some sample data I can test against, so I know what to make the hetzner data look like? If I know that, it's much easier to work on this, as I don't use AWS enough to have much in the way of real data to see what it's supposed to look like now.

Would the dummy demo data work in this case, or is that dummy data mimicing the billing based api , rather than usage data based approach?

Thanks

**Comment by mrchrisadams** [comment]
_I'm adding some notes here as I look in a bit more detail, before starting on this in earnest next week_

This here presumably is the shape of the data we need for each time series estimate to show stuff on a chart, that the front end is consuming:

https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/blob/trunk/packages/client/stub-server/mockData.json#L256-L331

So we'd need to be able to return an object that looks like a Services Estimate for each timestamp in a given time period:

```js
[   
 {
      "timestamp": "2021-03-25T00:00:00.000Z",
      "serviceEstimates": [
        {
          "cloudProvider": "AWS",
          "accountName": "aws account 0",
          "serviceName": "ebs",
          "kilowattHours": 0.9099318685999922,
          "co2e": 2,
          "cost": 2.0611247334106126,
          "region": "us-east-1",
          "usesAverageCPUConstant": false
        },
        {
          "cloudProvider": "AWS",
          "accountName": "aws account 0",
          "serviceName": "s3",
          "kilowattHours": 0.000986946262424086,
          "co2e": 1,
          "cost": 2.341122015683861,
          "region": "us-east-2",
          "usesAverageCPUConstant": false
        },
        {
          "cloudProvider": "AWS",
          "accountName": "aws account 3",
          "serviceName": "ec2",
          "kilowattHours": 50.176579830111365,
          "co2e": 3,
          "cost": 2.4154850519158613,
          "region": "us-west-1",
          "usesAverageCPUConstant": false
        },
      // snip - loads more of these objects, each containing a timestamp, and an array of ServiceEstimates
] // close the footprint array
```

Hezner provides block storage, so somewhat equivalent to AWS's EBS, and the equivalent to EC2 compute, in their form of compute.

So compute end up looking like this a sample Service Estimate along these lines:

```js
// 
  {
          "cloudProvider": "Hetzner",
          "accountName": "hetzner project name",
          "serviceName": "cloud",
          "kilowattHours": 50.176579830111365,
          "co2e": 3,
          "cost": 2.4154850519158613,
          "region": "hel1",
          "usesAverageCPUConstant": false
}
```

And storage might look like this:

```
{
          "cloudProvider": "Hetzner",
          "accountName": "hetzner project name",
          "serviceName": "volumes", // they call block storage volumes, i think
          "kilowattHours": 50.176579830111365,
          "co2e": 3,
          "cost": 2.4154850519158613,
          "region": "hel1",
          "usesAverageCPUConstant": false
}
```

https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/blob/trunk/packages/client/stub-server/mockData.json#L256-L331

If I'm able to output this, how much of the front end would be able to consume that? Would that just show in the chart, or would I get enough info for the emissions breakdown too?

**Comment by ccasher** [comment]
Hi @mrchrisadams, this all looks good. If you are able to provide that data that matches the stub server mock data, then it should populate all of the main charts on the landing page. As you can see in that file, there is also an emissions array, which would be used to color code the region breakdown chart if we are able to find the emissions factors for the new cloud provider regions. The emissions factors API provides the data returned from [this](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/blob/f7dbc44f351e2ac8295ef19040294ed964728130/packages/app/src/App.ts#L122) method
